### ISSUE TITLE
Codec set on MongoClient is overridden by DataStoreImpl

### ISSUE THREAD

**Issue by zenbones** [MorphiaOrg/morphia#1869]
This is with Morphia 2.2.2 and mongo driver 4.5.1

1    protected DatastoreImpl(MongoClient mongoClient, MapperOptions options, String dbName) {
2       MongoDatabase database = mongoClient.getDatabase(dbName);
3       this.mapper = new Mapper(this, database.getCodecRegistry(), options);
4        this.database = database.withCodecRegistry(mapper.getCodecRegistry());
5        this.mongoClient = mongoClient;
6
7        this.queryFactory = options.getQueryFactory();
8    }

The MongoClient handed in has my custom Codec, as dropping a breakpoint on using     mongoClient.getDatabase(dbName).getCodecRegistry().get(MyCustomClass.class) verifies. The database at line #2 still has the codec. However, at line #5 the database no longer has the codec. It's been overridden with database.withCodecRegistry(mapper.getCodecRegistry()).

The custom class for which I've registered a Codec happens to extend HashMap, and after line #4 the returned codec for that class is the default codec for Map, MorphiaMapCodec, but that's not what I want. How do I keep this from happening? Shouldn't the codec I specifically set on the MongoClient take precedent?

**Comment by evanchooly** [comment]
Is `MyCustomClass` annotated with `@Entity` or `@Embedded`?  If so, then Morphia will manage that class for you.  If you're providing your own codec, you don't need those annotations.  This might break if you're using them with `@Reference`.  It's not a tested scenario, honestly, but outside of that, leaving them unannotated should get you what you want.

**Comment by zenbones** [comment]
I appreciate the comment, thank you, but it fails to address the underlying concern. Whether I need the codec or not, someone will, and it's a feature, and it should work. I am suggesting that as currently implemented the code above is incorrect and should be considered a bug. My question really is 'Am I correct in this assumption?' Should it work as I am using the feature? If so, can I or a project maintainer correct this? If not, how should my use of the project change so that my codec gets called?

Now, to answer your question, MyCustomClass is neither @Entity nor @Embedded, nor can it be. It is, as I noted originally, an extension of HashMap whose keys are String and whose values are a complex type which I do not own, and therefore can not be annotated by me. Nor would their natural translation without annotation be something we want in the database for a number of reasons. I can write code to translate the values to and from Bson, thus the need for a custom codec. In the prior version of Morphia from which we're upgrading I had written a Converter, which worked perfectly. The project has decided to eliminate Converters because they were clear, and worked, and were easy enough. I assume there was some underlying reason for tossing an established part of the API. Fine, we've all been there. The problem persists and the replacement for Converter seems to be Codec. So I'm writing a Codec, but my Codec never gets called. Seems like a bug to me. Let's address that. How can I help?

**Comment by zenbones** [comment]
Doing some analysis, it looks like ProvidersCodecRegistry conatins this simple code...

```
    public <T> Codec<T> get(final Class<T> clazz, final CodecRegistry registry) {
        for (CodecProvider provider : codecProviders) {
            Codec<T> codec = provider.get(clazz, registry);
            if (codec != null) {
                return codec;
            }
        }
        return null;
    }
```
...which will return the first matching codec found. That's reasonable. However, dev.morphia.mapping.Mapper gets constructed like this...
   ```
 public Mapper(Datastore datastore, CodecRegistry codecRegistry, MapperOptions options) {
        this.datastore = datastore;
        this.options = options;
        morphiaCodecProvider = new MorphiaCodecProvider(this, datastore);
        discriminatorLookup = new DiscriminatorLookup(options.getClassLoader());

        this.codecRegistry = fromProviders(new MorphiaTypesCodecProvider(this),
            new PrimitiveCodecRegistry(codecRegistry),
            new EnumCodecProvider(),
            new AggregationCodecProvider(this),
            morphiaCodecProvider,
            codecRegistry);
    }
```
...which puts the codec from MongoClient last in the list to be checked, in the fallback position, unless, I think, my type override happen to be primitive (which, in this case, it's not). As the codec in MongoClient is the one the user of the project has access to, I believe this order is clearly incorrect. As is, any codec written for a type that matches a default codec, or written for a type whose super types or implemented interfaces match a default codec (in other words if the type is coercible into a type that matches a default code), will never be called. That makes the entire system somewhat useless, but the fix is easy. Just move the code registry passed in the constructor first in the list...
 ```
       this.codecRegistry = fromProviders(new MorphiaTypesCodecProvider(this),
            codecRegistry,
            new EnumCodecProvider(),
            new AggregationCodecProvider(this),
            morphiaCodecProvider);
```
...which is just to assume that the person using the project actually knew what they wanted. If this is a breaking change, and it probably is as it changes assumptions, then we need to add a flag to the MongoClient construction, that gets passed all the way through to the Mapper construction, that allows the user to declare that they want their codecs checked first.

**Comment by zenbones** [comment]
Opened a pull request on master with the requisite changes.

**Comment by evanchooly** [comment]
Simply moving the driver specified codecs to the front breaks other things (e.g., handling of lists, iirc, because morphia does a bit more with them).  If you have registered your codec with the driver, but don't have your type annotated, it should find your codec and use it.  If you could put together a [reproducer](/morphiaorg/reproducer) I'd be happy to take a look and see what's going on.

**Comment by evanchooly** [comment]
It's possible that the `MorphiaTypesCodecProvider` is the one to blame here but I'd have to look at a reproducer to see exactly what's happening in your case.

**Comment by evanchooly** [comment]
fwiw, [this issue](https://github.com/MorphiaOrg/morphia/issues/1842) directly relates to yours and would give you the functionality you need.

**Comment by zenbones** [comment]
Don't know which provider is the blocker, but having a method on client which allows custom codecs to be set, with examples of how to do so, only to have them ignored in practice, *is* a bug and should be resolved. I get that resolution must be in due time. If there are codecs required, such as for list handling, that *must* be ahead of custom codecs, then those should be broken out from the rest and ordered properly such that user defined codecs are still *always* consulted first. Even if a user has defined a codec for a list, for example. It should be first, because ignoring a user defined codec must be, a priori, wrong.

Fixing this may need a deeper dive into the codecs required by Morphia and how those are applied to the system versus user defined codecs. Maybe there's work to be done on how codec registries work, or on class resolution. I don't know. But the end result should be that user defined codecs that are applied via...

`MongoClientSettings.builder().codecRegistry()`

are engaged on the classes for which they are defined, or the method above should be removed for a method which behaves as expected.

**Comment by zenbones** [comment]
I ay try reproducer, but a reproduction is as easy as...

```
public class MyClass extends HashMap<String, String> {}

@Entity
public class MyPersisted {

  @Property
  private MyClass myClass;
}

public class MyClassCodec implements Codec<MyClass> {
  ... methods can do nothing
}
```

Add the codec to a client using...

`CodecRegistries.fromRegistries(CodecRegistries.fromCodecs(new MyClassCodec()), MongoClientSettings.getDefaultCodecRegistry());`

And see that the codec is never called and MorphiaMapCodec is used instead.

**Comment by evanchooly** [comment]
Yeah, that's as I suspected.  Morphia's codec is getting invoked because your type extends `Map`.  In 2.2, there's not really a way of delineating the distinction you're trying to make but the linked issue in 2.3 will do so.

**Comment by zenbones** [comment]
Do you have a link to the code change in 2.3 which corrects this? I should probably be able to find it but not sure how to navigate GitHub correctly. I've come this far, so I thought I'd do a code review and put my 2 cents in before this goes public, if you don't mind another set of eyes.

**Comment by evanchooly** [comment]
I haven't implemented it yet but it's next on my list.

**Comment by zenbones** [comment]
Can I suggest that the codec registry set on `MongoClientSettings.builder().codecRegistry()` not replace the default codec registy on the MongoClient, but rather be added internally as a separate custom codec registry. Then, in DataStoreImpl rather than trying to parse out bits of the default codecs and push the rest later, the custom codec registry can be placed first knowing that it contains only user added codecs which should always be checked first, then you can pull out the primitive codecs from the default codec registry and push the rest of the defaults later down the line, as is being done  now, and that should be safe for both old and new usages. Something like...

```
        morphiaCodecProviders.add(new MorphiaCodecProvider(this));

        CodecRegistry codecRegistry = database.getCodecRegistry();
        List<CodecProvider> providers = new ArrayList<>();

          // Here are only custom codecs added by the user in the constructor to MongoClient
           providers.add(customCodecRegistry);

          // Everything below is in the same order it's always had, so the only codecs shifted to the top are the custom
          // codecs added by the user 

           providers.addAll(List.of(new MorphiaTypesCodecProvider(this),
           // These are the default codecs in MongoClient
            new PrimitiveCodecRegistry(defaultCodecRegistry),
            new EnumCodecProvider(),
            new AggregationCodecProvider(this)));

        providers.addAll(morphiaCodecProviders);
        // These are the default codecs in MongoClient, again
        providers.add(defaultCodecRegistry);
        this.codecRegistry = fromProviders(providers);
```

**Comment by evanchooly** [comment]
fixed by #1842
